RFD 51: X11 forwarding (#10009)
gravitational/teleport: rfd/0051-x11-forwarding.md

---
authors: Brian Joerger (bjoerger@goteleport.com)
state: implemented
---

# RFD 51 - X11 Forwarding

## What

X11 forwarding makes it possible to open a graphical application on a remote
machine and but forward its graphical (X) data to the client's local display.  

> SSH rfc [4254](https://www.rfc-editor.org/rfc/rfc4254#section-6.3) details the

Reviewer: nit:
```suggestion
[RFC4254](https://www.rfc-editor.org/rfc/rfc4254#section-6.3) details the
```

protocol for supporting X11 forwarding in an SSH server/client.

## Why

X11 forwarding is used by some SSH users who are switching over to Teleport.
Currently, we only support X11 forwarding between an OpenSSH client and server

> in proxy recording mode. As user's move more of their remote fleet to Teleport,

Reviewer: ```suggestion
in proxy recording mode. As users move more of their fleet to Teleport,
```

2. Server receives request and either approves or denies the request.
3. If the server approves, it will open an XServer proxy listener to catch any
   XServer requests within the SSH session.
4. Upon receiving an XServer request on the proxy listener, the server sends an
   `x11` channel request to the client.
5. If the client accepts, the client will start X11 forwarding between the x11
   channel and the client's local display - `$DISPLAY`.
6. Server begins X11 forwarding between the x11 channel and XServer proxy connection.
7. The XServer request is now connected to the client's local display. For example, if
   `xeyes` was run, then the program would be running on the server, but any graphical
   input/output would the client's local XServer.

### X Server Proxy

In step 3 above, the SSH server opens an X Server Proxy. More specifically, the server
starts listening on a unix socket so that it can accept requests to the corresponding
display. 

First, the server has to find an open display socket. A display's unix socket can be 
determined by its display number - `/tmp/.X11-unix/X<display_number>`. So we start
searching for an open display number between 10 and 1009.

> Note: We use the same display offset (10) and max displays (1000) as OpenSSH.

Reviewer: Why is an offset necessary? Why not start looking at display 0?

Author: To avoid overlapping with local displays which start counting up from 0. It's not strictly necessary, but for consistency/clarity we add the offset to separate local and proxy displays. This way `ls /tmp/.X11-unix` might output `X0 X1 X2 X10 X11 X12 X13`, and it's clear where each unix socket originated. Adding a shorter explanation to the RFD.

was the fake data we created. So before forwarding the X request to the X Server, the
SSH client replaces the fake data with real X authorization data. This data will either
be trusted xauth data, or untrusted xauth data, depending on what the client requested.

##### Untrusted X11 Forwarding

Untrusted X authorization data must be generated through `xauth` in order to be recognized
by the X Server. To do this, the client runs `xauth generate <$DISPLAY> untrusted <timeout?>`.
Any X Client request using this untrusted token will be be subject to the X Security extension,
granting it fewer privileges than usual. 

##### Trusted X11 Forwarding

In trusted X11 forwarding, we can rely on the client's local environment. If we fill in 
the X authorization packet in the request with a random cookie, the X Server will just
ignore it and connect as if it received the request locally.

### tsh UX

`tsh ssh -X` or `tsh ssh -Y` can be used to request a session with `trusted` X11 forwarding.
However, if `tsh ssh -X` is used with the `-oForwardX11Trusted=yes` flag, it will instead
request `untrusted` forwarding, while `tsh ssh -Y` would still request `trusted`.

> This may seem unintuitive, and it is, but this is the same UX as OpenSSH client. In OpenSSH,

Reviewer: This doesn't seem like a good enough reason to me.

I wonder if it would be worth creating a more intuitive UX, and also supporting the openssh flags for compatibility.

Author: Agreed, updated the UX and added an `deviations from openssh` section to explain the change.

extension) is an opt in extension of the X Protocol with negative performance and compatibility
implications.

### X Security extension

The X Security extension was created about 10 years after the 11th version of X. The
extension introduced `untrusted` tokens which could be used to grant fewer X privileges
to the user of the token. It also introduced the ability to set a timeout for the token.

In contrast, normal `trusted` tokens give full unmitigated X privileges, allowing for action
such as screenshots, monitoring peripherals, and many other actions one would expect to
perform on local machines. In the case of X11 forwarding however, this means that the
client is providing the SSH server unmitigated access to its local X Server. If the server
becomes compromised or another user has access to your remote user, then the Client's
local XServer could be subject to attacks such as keystroke monitoring to capture a password.

Unfortunately, since the X Security extension was created as an afterthought, it is not
built into the X protocol and has some serious performance drawbacks. The performance 

> drop is somewhere in the ballpark of 10x. Additionally, many X Client's don't support

Reviewer: ```suggestion
drop is somewhere in the ballpark of 10x. Additionally, many X Clients don't support
```

> the X Security Extension properly and may crash immediately or at some point during 

Reviewer: Can we link to any issues or examples here?

Author: I added a general example but couldn't find anything specific. This may be a much smaller issue than it was 20 years ago, hard to tell to be honest.

release. On top of the security benefits being confusing and unclear to many users, using the
extension initially led to many X programs to perform poorly or even crash, as mentioned in the
previous section.

It seems that this UX decision is ultimately the result of trying to include an incompatible
yet important security fix without having any effect on existing users of X11 forwarding.
In other words the X Security Extension is an opt-in feature in OpenSSH X11 forwarding as a
result of its issues.

Two decades later, I believe we have no reason to follow the same UX and should instead learn
from the mistakes of OpenSSH. As a security centered product, the security implications of a 
user's actions should be as clear and unambiguous as possible. 

From the get-go, `tsh ssh -X` and `tsh ssh -Y` will always lead  to `untrusted` or `trusted`
forwarding respectively. If a user runs into an X program which still doesn't support the X 
Security Extension, I would prefer that they experience the crash, come to customer support
or other sources to discover the reason, and personally decide whether to switch to `trusted`
forwarding or keep their X Server secure.

> This does leave us with the unfortunate fact that the UX of `tsh` and `ssh` will be different

Reviewer: If this became a problem, we could technically add some code to detect if `tsh` is aliased and being invoked as `ssh`, and supporting the `ssh` flags only in that case.

Right?
